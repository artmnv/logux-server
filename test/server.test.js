'use strict'

const spawn = require('child_process').spawn
const path = require('path')

const Server = require('../server')

const DATE = /\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d/g

function wait (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

let started

function start (name, args) {
  return new Promise(resolve => {
    started = spawn(path.join(__dirname, '/servers/', name), args)
    let running = false
    function callback () {
      if (!running) {
        running = true
        resolve()
      }
    }
    started.stdout.on('data', callback)
    started.stderr.on('data', callback)
  })
}

function test (name, args) {
  return new Promise(resolve => {
    let out = ''
    const server = spawn(path.join(__dirname, '/servers/', name), args)
    server.stdout.on('data', chank => {
      out += chank
    })
    server.stderr.on('data', chank => {
      out += chank
    })
    server.on('close', exitCode => {
      let fixed = out.replace(DATE, '1970-01-01 00:00:00')
                     .replace(/PID:(\s+)\d+/, 'PID:$121384')
                     .replace(/"pid":\d+,/g, '"pid":21384,')
                     .replace(
                       /"hostname":"[-.\w]+",/g,
                       '"hostname":"MacBook-Pro",'
                     )
                     .replace(
                       /"time":"[-.:\dTZ]+",/g,
                       '"time":"1970-01-01T00:00:00.000Z",'
                     )
      fixed = fixed.replace(/\r\v/g, '\n')
      resolve([fixed, exitCode])
    })
    wait(500).then(() => {
      server.kill('SIGINT')
    })
  })
}

function checkOut (name, args) {
  return test(name, args).then(result => {
    const out = result[0]
    const exit = result[1]

    if (exit !== 0) {
      console.error(`${ test } fall with:\n${ out }`)
    }
    expect(exit).toEqual(0)
    expect(out).toMatchSnapshot()
  })
}

function checkError (name, args) {
  return test(name, args).then(result => {
    const out = result[0]
    const exit = result[1]
    expect(exit).toEqual(1)
    expect(out).toMatchSnapshot()
  })
}

const originEnv = process.env.NODE_ENV
afterEach(() => {
  process.env.NODE_ENV = originEnv
  delete process.env.LOGUX_PORT
  if (started) {
    started.kill('SIGINT')
    started = undefined
  }
})

it('uses cli args for options', () => {
  const options = Server.prototype.loadOptions({
    argv: ['', '--port', '31337', '--host', '192.168.1.1'],
    env: { }
  })

  expect(options.host).toEqual('192.168.1.1')
  expect(options.port).toEqual(31337)
  expect(options.cert).toBeUndefined()
  expect(options.key).toBeUndefined()
})

it('uses env for options', () => {
  const options = Server.prototype.loadOptions({
    argv: [],
    env: { LOGUX_HOST: '127.0.1.1', LOGUX_PORT: 31337 }
  })

  expect(options.host).toEqual('127.0.1.1')
  expect(options.port).toEqual(31337)
})

it('uses combined options', () => {
  const options = Server.prototype.loadOptions({
    env: { LOGUX_CERT: './cert.pem' },
    argv: ['', '--key', './key.pem']
  }, { port: 31337 })

  expect(options.port).toEqual(31337)
  expect(options.cert).toEqual('./cert.pem')
  expect(options.key).toEqual('./key.pem')
})

it('uses arg, env, defaults options in given priority', () => {
  const options1 = Server.prototype.loadOptions({
    argv: ['', '--port', '31337'],
    env: { LOGUX_PORT: 21337 }
  }, { port: 11337 })
  const options2 = Server.prototype.loadOptions({
    argv: [],
    env: { LOGUX_PORT: 21337 }
  }, { port: 11337 })

  expect(options1.port).toEqual(31337)
  expect(options2.port).toEqual(21337)
})

it('destroys everything on exit', () => checkOut('destroy.js'))

it('reports unbind', () => test('unbind.js').then(result => {
  expect(result[0]).toMatchSnapshot()
}))

it('shows uncatch errors', () => checkError('throw.js'))

it('shows uncatch rejects', () => checkError('uncatch.js'))

it('use environment variables for config', () => {
  process.env.LOGUX_PORT = 31337
  return checkOut('options.js')
})

it('shows help', () => checkOut('options.js', ['', '--help']))

it('shows help about port in use', () => start('eaddrinuse.js')
  .then(() => test('eaddrinuse.js')).then(result => {
    expect(result[0]).toMatchSnapshot()
  }))

it('shows help about privileged port', () => checkError('eacces.js'))

it('reports to bunyan log', () => checkOut('bunyan.js'))

it('reports to custom bunyan log', () => checkOut('bunyan-custom.js'))
