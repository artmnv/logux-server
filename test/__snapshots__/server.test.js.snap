// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destroys everything on exit 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.1.1
        Node ID:      server:r1CJmycQW
        Environment:  test
        Subprotocol:  1.0.0
        Supports:     1.x
        Listen:       ws://127.0.0.1:2000

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

 Custom destroy task finished
"
`;

exports[`disables colors for constructor errors 1`] = `
" ERROR  Missed client subprotocol requirements at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`reports to bunyan log 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.1.1\\",\\"nodeId\\":\\"server:r1CJmycQW\\",\\"environment\\":\\"test\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:2000\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`reports to custom bunyan log 1`] = `
"{\\"name\\":\\"logux-server-custom\\",\\"customProp\\":\\"42\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.1.1\\",\\"nodeId\\":\\"server:r1CJmycQW\\",\\"environment\\":\\"test\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:2000\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server-custom\\",\\"customProp\\":\\"42\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`reports unbind 1`] = `
" INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`shows help 1`] = `
"Options:
  -h, --host      Host to bind server.                                  [string]
  -p, --port      Port to bind server                                   [number]
  -k, --key       Path to SSL key                                       [string]
  -c, --cert      Path to SSL certificate                               [string]
  -r, --reporter  Reporter type            [string] [choices: \\"human\\", \\"bunyan\\"]
  --help          Show help                                            [boolean]

Examples:
  test/servers/options.js --port 31337 --host 127.0.0.1
  LOGUX_PORT=1337 test/servers/options.js

Environment variables:
  LOGUX_HOST, LOGUX_PORT, LOGUX_KEY, LOGUX_CERT, LOGUX_REPORTER

"
`;

exports[`shows help about missed option 1`] = `
" ERROR  Missed client subprotocol requirements at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`shows help about port in use 1`] = `
" ERROR  Port :2001 already in use at 1970-01-01 00:00:00
        Another Logux server or other app already running on this port
        Maybe you didn’t not stop server from other project
        or previous version of this server was not killed

"
`;

exports[`shows help about privileged port 1`] = `
" ERROR  You are not allowed to run server on this port at 1970-01-01 00:00:00
        Try to change user (e.g. root) or use port >= 1024

"
`;

exports[`shows help about unknown option 1`] = `
" ERROR  Unknown option \`support\` in server constructor at 1970-01-01 00:00:00
        Maybe there is a mistake in option name or this version
        of Logux Server doesn’t support this option

"
`;

exports[`shows uncatch errors 1`] = `
" ERROR  Test Error at 1970-01-01 00:00:00
        fake stacktrace

Error event: Test Error
 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`shows uncatch rejects 1`] = `
" ERROR  Test Error at 1970-01-01 00:00:00
        fake stacktrace

Error event: Test Error
 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`use environment variables for config 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.1.1\\",\\"nodeId\\":\\"server:r1CJmycQW\\",\\"environment\\":\\"test\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:31337\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`uses reporter param 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.1.1\\",\\"nodeId\\":\\"server:r1CJmycQW\\",\\"environment\\":\\"test\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:1337\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`uses reporter param for constructor errors 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":50,\\"hint\\":[\\"Check server constructor and Logux Server documentation\\"],\\"msg\\":\\"Missed client subprotocol requirements\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;
